src/python/pants/jvm/resolve/jvm_tool.py

```diff
@@ -161,5 +162,45 @@ async def gather_coordinates_for_jvm_lockfile(
     return ArtifactRequirements(requirements)
 
 
+@frozen_after_init
+@dataclass
```

Comment:
You can get rid of the custom `__hash__` with this

```suggestion
@dataclass(unsafe_hash=True)
```